File: cpp/src/api_key_context.cpp
// Copyright (c), CommunityLogiq Software

#include "ulsdk/api_key_context.h"

#include <curl/curl.h>

#include <algorithm>
#include <map>
#include <sstream>
#include <string>

#include "ulsdk/keys.h"
#include "ulsdk/ulsdk.h"

namespace ul {
static const char* REQUEST_TYPE = "ul1_request";
static const char* SIGNATURE_V1 = "UL1-ED25519";
static const char* HEADER_AUTHORIZATION = "authorization";
static const char* HEADER_X_UL_DATE = "x-ul-date";
static const char* HEADER_CONTENT_TYPE = "content-type";
static const char* SIGNED_HEADERS[] = {HEADER_X_UL_DATE};

class CurlSlistHandle {
    struct curl_slist* _slist;

public:
    CurlSlistHandle() : _slist(nullptr) {}

    ~CurlSlistHandle() {
        curl_slist_free_all(_slist);
    }

    CurlSlistHandle(const CurlSlistHandle&) = delete;

    struct curl_slist* get() {
        return _slist;
    }

    void set(struct curl_slist* slist) {
        _slist = slist;
    }
};

class CurlHandle {
    CURL* _curl;
    char* _error_buffer;

public:
    CurlHandle() {
        _error_buffer = new char[CURL_ERROR_SIZE];
        _curl = curl_easy_init();
        curl_easy_setopt(_curl, CURLOPT_ERRORBUFFER, _error_buffer);
    }

    ~CurlHandle() {
        curl_easy_cleanup(_curl);
        delete[] _error_buffer;
    }

    CurlHandle(const CurlHandle&) = delete;

    bool ok() const {
        return _curl != nullptr;
    }

    operator CURL*() const {
        return _curl;
    }

    const char* error_buffer() const {
        return _error_buffer;
    }
};

static Result<std::vector<uint8_t>>
raise_for_status(CURL* curl, std::vector<uint8_t> response) {
    long code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &code);
    if (code >= 400) {
        return Result<std::vector<uint8_t>>(
            Error((int)code, std::string(response.begin(), response.end()))
        );
    }

    return Result<std::vector<uint8_t>>(response);
}

static std::string add_params(
    CURL* curl,
    const std::string& url,
    const std::vector<std::pair<std::string, std::string>>& params
) {
    std::string result = url;
    if (params.empty()) {
        return result;
    }

    result.push_back('?');
    for (const auto& pair : params) {
        if (result.back() != '?') {
            result.push_back('&');
        }
        result += pair.first;
        result.push_back('=');
        const char* encoded =
            curl_easy_escape(curl, pair.second.c_str(), pair.second.size());
        result += encoded;
        curl_free((void*)encoded);
    }

    return result;
}

static size_t
write_callback(void* ptr, size_t size, size_t nmemb, void* userdata) {
    std::vector<uint8_t>& response = *(std::vector<uint8_t>*)userdata;
    const size_t total = size * nmemb;
    response.insert(response.end(), (uint8_t*)ptr, (uint8_t*)ptr + total);
    return total;
}

static std::string to_upper(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(), ::toupper);
    return result;
}

static std::string to_lower(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(), ::tolower);
    return result;
}

static std::string canonicalize_path(const std::string& path) {
    if (path.empty()) {
        return "/";
    }

    if (path[0] != '/') {
        return std::string("/") + path;
    }

    return path;
}

static Result<std::string> canonicalize_query_string(
    CURL* curl,
    const std::vector<std::pair<std::string, std::string>>& query
) {
    if (query.empty()) {
        return Result<std::string>(std::string());
    }

    std::vector<std::pair<std::string, std::string>> query_copy = query;
    std::sort(
        query_copy.begin(),
        query_copy.end(),
        [](const auto& a, const auto& b) { return a.first < b.first; }
    );

    std::string result;
    for (const auto& pair : query) {
        if (pair.first == "X-UL-Signature") {
            continue;
        }

        const char* encoded =
            curl_easy_escape(curl, pair.second.c_str(), pair.second.size());
        if (encoded == nullptr) {
            return Result<std::string>(Error("Failed to encode query parameter")
            );
        }
        const std::string value(encoded);
        curl_free((void*)encoded);

        if (!result.empty()) {
            result.push_back('&');
        }
        result += pair.first;
        result.push_back('=');
        result += value;
    }

    return Result<std::string>(result);
}

static Result<std::string> canonicalize_headers(
    const std::vector<std::string>& signed_headers,
    const std::vector<std::pair<std::string, std::string>>& headers
) {
    std::vector<std::string> sorted_signed_headers;
    for (const auto& header : signed_headers) {
        sorted_signed_headers.push_back(to_lower(header));
    }
    std::sort(sorted_signed_headers.begin(), sorted_signed_headers.end());

    std::string canonical_header_parts;
    for (const auto& header : sorted_signed_headers) {
        for (const auto& pair : headers) {
            if (to_lower(pair.first) == header) {
                if (!canonical_header_parts.empty()) {
                    canonical_header_parts.push_back('\n');
                }

                canonical_header_parts += header;
                canonical_header_parts.push_back(':');
                canonical_header_parts += pair.second;
            }
        }
    }

    return Result<std::string>(canonical_header_parts);
}

static std::string hash(size_t len, const uint8_t* data) {
    uint8_t hash_out[crypto_hash_sha256_BYTES];
    crypto_hash_sha256(hash_out, data, len);

    return hex(crypto_hash_sha256_BYTES, hash_out);
}

static Result<std::pair<std::string, std::string>> canonicalize_request(
    const CurlHandle& curl,
    const std::string& method,
    const std::string& path,
    const std::vector<std::pair<std::string, std::string>>& query,
    const std::vector<std::pair<std::string, std::string>>& headers,
    const std::vector<std::string>& signed_headers,
    const std::vector<uint8_t>& body
) {
    const std::string canonical_path = canonicalize_path(path);

    const Result<std::string> canonical_query_res =
        canonicalize_query_string(curl, query);
    if (std::holds_alternative<Error>(canonical_query_res)) {
        return Result<std::pair<std::string, std::string>>(
            std::get<Error>(canonical_query_res)
        );
    }
    const std::string canonical_query_string =
        std::get<std::string>(canonical_query_res);

    const Result<std::string> canonical_headers_res =
        canonicalize_headers(signed_headers, headers);
    if (std::holds_alternative<Error>(canonical_headers_res)) {
        return Result<std::pair<std::string, std::string>>(
            std::get<Error>(canonical_headers_res)
        );
    }
    const std::string canonical_headers =
        std::get<std::string>(canonical_headers_res);

    std::string signed_headers_str;
    for (const auto& header : signed_headers) {
        if (!signed_headers_str.empty()) {
            signed_headers_str.push_back(';');
        }
        signed_headers_str += header;
    }

    std::stringstream ss;
    ss << to_upper(method) << '\n'
       << canonical_path << '\n'
       << canonical_query_string << '\n'
       << canonical_headers << '\n'
       << signed_headers_str << '\n'
       << hash(body.size(), body.data());

    const std::string s = ss.str();
    const std::string h = hash(s.size(), (const uint8_t*)s.c_str());

    return Result<std::pair<std::string, std::string>>(
        std::make_pair(h, signed_headers_str)
    );
}

static Result<std::string> generate_auth_header(
    const CurlHandle& curl,
    const Key& key,
    const std::string& method,
    const std::string& path,
    const std::vector<std::pair<std::string, std::string>>& query,
    std::vector<std::pair<std::string, std::string>>& headers,
    const std::vector<uint8_t>& body
) {
    const unsigned long ts = time(nullptr);

    headers.push_back(std::make_pair(HEADER_X_UL_DATE, std::to_string(ts)));

    std::vector<std::string> signed_headers;
    for (const auto& header : SIGNED_HEADERS) {
        signed_headers.push_back(header);
    }

    const Result<std::pair<std::string, std::string>> canonical_request_res =
        canonicalize_request(
            curl,
            method,
            path,
            query,
            headers,
            signed_headers,
            body
        );

    if (std::holds_alternative<Error>(canonical_request_res)) {
        return Result<std::string>(std::get<Error>(canonical_request_res));
    }

    const std::pair<std::string, std::string> canonical_request =
        std::get<std::pair<std::string, std::string>>(canonical_request_res);
    const std::string request_hash = canonical_request.first;
    const std::string signed_header_string = canonical_request.second;

    std::stringstream scope;
    scope << key.user_id_.to_string() << '/' << ts << '/'
          << to_string(key.region_) << '/' << REQUEST_TYPE;

    std::stringstream signing_string_stream;
    signing_string_stream << SIGNATURE_V1 << '\n'
                          << scope.str() << '\n'
                          << request_hash;
    const std::string signing_string = signing_string_stream.str();

    unsigned char signature[crypto_sign_BYTES];
    crypto_sign_detached(
        signature,
        nullptr,
        (const uint8_t*)signing_string.c_str(),
        signing_string.size(),
        key.secret_key_.data()
    );
    const std::string hex_signature = hex(crypto_sign_BYTES, signature);

    std::stringstream header;
    header << SIGNATURE_V1 << " Credential=" << key.access_key_ << '/'
           << scope.str() << ", SignedHeaders=" << signed_header_string
           << ", Signature=" << hex_signature;

    return Result<std::string>(header.str());
}

ApiKeyContext::ApiKeyContext(const Key& key, const Environment& environment) :
    environment_(environment),
    key_(key) {}

ApiKeyContext::~ApiKeyContext() {}

Region ApiKeyContext::region() const {
    return key_.region_;
}

Environment ApiKeyContext::environment() const {
    return environment_;
}

Result<std::vector<uint8_t>> ApiKeyContext::get(
    const std::string& path,
    const std::map<std::string, std::string>& params_map,
    const std::map<std::string, std::string>& headers_map
) const {
    const Result<std::string> endpoint =
        get_endpoint(key_.region_, environment_, path);
    if (std::holds_alternative<Error>(endpoint)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(endpoint));
    }

    CurlHandle curl;
    if (!curl.ok()) {
        return Result<std::vector<uint8_t>>(Error("Failed to initialize curl"));
    }

    const auto params = std::vector<std::pair<std::string, std::string>>(
        params_map.begin(),
        params_map.end()
    );
    auto headers = std::vector<std::pair<std::string, std::string>>(
        headers_map.begin(),
        headers_map.end()
    );

    const Result<std::string> auth_header_res = generate_auth_header(
        curl,
        key_,
        "GET",
        path,
        params,
        headers,
        std::vector<uint8_t>()
    );
    if (std::holds_alternative<Error>(auth_header_res)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(auth_header_res));
    }

    const std::string auth_header = std::get<std::string>(auth_header_res);

    CurlSlistHandle slist;
    for (const auto& pair : headers) {
        std::stringstream header_stream;
        header_stream << pair.first << ": " << pair.second;
        const std::string header = header_stream.str();

        slist.set(curl_slist_append(slist.get(), header.c_str()));
    }
    std::stringstream auth_header_stream;
    auth_header_stream << HEADER_AUTHORIZATION << ": " << auth_header;
    const std::string auth_header_str = auth_header_stream.str();
    slist.set(curl_slist_append(slist.get(), auth_header_str.c_str()));

    std::vector<uint8_t> response;
    const std::string url =
        add_params(curl, std::get<std::string>(endpoint), params);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, slist.get());
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 0L);
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);

    if (curl_easy_perform(curl) != CURLE_OK) {
        return Result<std::vector<uint8_t>>(Error(curl.error_buffer()));
    }

    return raise_for_status(curl, response);
}

Result<std::vector<uint8_t>> ApiKeyContext::put(
    const std::string& path,
    const std::vector<uint8_t>& data,
    const std::string& mimetype,
    const std::map<std::string, std::string>& params_map,
    const std::map<std::string, std::string>& headers_map
) const {
    const Result<std::string> endpoint =
        get_endpoint(key_.region_, environment_, path);
    if (std::holds_alternative<Error>(endpoint)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(endpoint));
    }

    CurlHandle curl;
    if (!curl.ok()) {
        return Result<std::vector<uint8_t>>(Error("Failed to initialize curl"));
    }

    const auto params = std::vector<std::pair<std::string, std::string>>(
        params_map.begin(),
        params_map.end()
    );
    auto headers = std::vector<std::pair<std::string, std::string>>(
        headers_map.begin(),
        headers_map.end()
    );

    const Result<std::string> auth_header_res =
        generate_auth_header(curl, key_, "PUT", path, params, headers, data);
    if (std::holds_alternative<Error>(auth_header_res)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(auth_header_res));
    }

    const std::string auth_header = std::get<std::string>(auth_header_res);

    CurlSlistHandle slist;
    for (const auto& pair : headers) {
        std::stringstream header_stream;
        header_stream << pair.first << ": " << pair.second;
        const std::string header = header_stream.str();

        slist.set(curl_slist_append(slist.get(), header.c_str()));
    }

    std::stringstream auth_header_stream;
    auth_header_stream << HEADER_AUTHORIZATION << ": " << auth_header;
    const std::string auth_header_str = auth_header_stream.str();
    slist.set(curl_slist_append(slist.get(), auth_header_str.c_str()));

    std::stringstream content_type_stream;
    content_type_stream << HEADER_CONTENT_TYPE << ": " << mimetype;
    const std::string content_type = content_type_stream.str();
    slist.set(curl_slist_append(slist.get(), content_type.c_str()));

    std::vector<uint8_t> response;
    const std::string url =
        add_params(curl, std::get<std::string>(endpoint), params);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, slist.get());
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 0L);
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data.data());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, data.size());
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);

    if (curl_easy_perform(curl) != CURLE_OK) {
        return Result<std::vector<uint8_t>>(Error(curl.error_buffer()));
    }

    return raise_for_status(curl, response);
}

Result<std::vector<uint8_t>> ApiKeyContext::post(
    const std::string& path,
    const std::vector<uint8_t>& data,
    const std::string& mimetype,
    const std::map<std::string, std::string>& params_map,
    const std::map<std::string, std::string>& headers_map
) const {
    const Result<std::string> endpoint =
        get_endpoint(key_.region_, environment_, path);
    if (std::holds_alternative<Error>(endpoint)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(endpoint));
    }

    CurlHandle curl;
    if (!curl.ok()) {
        return Result<std::vector<uint8_t>>(Error("Failed to initialize curl"));
    }

    const auto params = std::vector<std::pair<std::string, std::string>>(
        params_map.begin(),
        params_map.end()
    );
    auto headers = std::vector<std::pair<std::string, std::string>>(
        headers_map.begin(),
        headers_map.end()
    );

    const Result<std::string> auth_header_res =
        generate_auth_header(curl, key_, "POST", path, params, headers, data);
    if (std::holds_alternative<Error>(auth_header_res)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(auth_header_res));
    }

    const std::string auth_header = std::get<std::string>(auth_header_res);

    CurlSlistHandle slist;
    for (const auto& pair : headers) {
        std::stringstream header_stream;
        header_stream << pair.first << ": " << pair.second;
        const std::string header = header_stream.str();

        slist.set(curl_slist_append(slist.get(), header.c_str()));
    }

    std::stringstream auth_header_stream;
    auth_header_stream << HEADER_AUTHORIZATION << ": " << auth_header;
    const std::string auth_header_str = auth_header_stream.str();
    slist.set(curl_slist_append(slist.get(), auth_header_str.c_str()));

    std::stringstream content_type_stream;
    content_type_stream << HEADER_CONTENT_TYPE << ": " << mimetype;
    const std::string content_type = content_type_stream.str();
    slist.set(curl_slist_append(slist.get(), content_type.c_str()));

    std::vector<uint8_t> response;
    const std::string url =
        add_params(curl, std::get<std::string>(endpoint), params);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, slist.get());
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 0L);
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data.data());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, data.size());
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);

    if (curl_easy_perform(curl) != CURLE_OK) {
        return Result<std::vector<uint8_t>>(Error(curl.error_buffer()));
    }

    return raise_for_status(curl, response);
}

Result<std::vector<uint8_t>> ApiKeyContext::upload(
    const std::string& path,
    const std::vector<File>& files
) const {
    // This implementation manually constructs the request body as the entire
    // body is needed in order to calculate the signature, which I don't think
    // we can do with libcurl's multipart form data support.
    const unsigned long ts = time(nullptr);
    std::stringstream boundary_stream;
    boundary_stream << "--UL1-multipart-" << ts;
    const std::string boundary = boundary_stream.str();
    const std::string crlf = "\r\n";

    std::vector<uint8_t> body;
    body.insert(body.end(), boundary.begin(), boundary.end());
    body.insert(body.end(), crlf.begin(), crlf.end());

    for (const auto& file : files) {
        if (!body.empty()) {
            body.insert(body.end(), boundary.begin(), boundary.end());
            body.insert(body.end(), crlf.begin(), crlf.end());
        }

        std::stringstream disposition_header;
        disposition_header << "Content-Disposition: form-data; name=\""
                           << file.name << "\"; filename=\"" << file.name
                           << "\"\r\n";
        disposition_header << "Content-Type:" << file.mimetype << "\r\n\r\n";
        const std::string header = disposition_header.str();

        body.insert(body.end(), header.begin(), header.end());
        body.insert(body.end(), file.data.begin(), file.data.end());
    }
    body.insert(body.end(), boundary.begin(), boundary.end());
    body.push_back((uint8_t)'-');
    body.push_back((uint8_t)'-');

    std::stringstream mime_type;
    mime_type << "multipart/form-data; boundary=\"" << boundary << "\"";

    return post(path, body, mime_type.str(), {}, {});
}

Result<std::vector<uint8_t>> ApiKeyContext::del(
    const std::string& path,
    const std::map<std::string, std::string>& params_map,
    const std::map<std::string, std::string>& headers_map
) const {
    const Result<std::string> endpoint =
        get_endpoint(key_.region_, environment_, path);
    if (std::holds_alternative<Error>(endpoint)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(endpoint));
    }

    CurlHandle curl;
    if (!curl.ok()) {
        return Result<std::vector<uint8_t>>(Error("Failed to initialize curl"));
    }

    const auto params = std::vector<std::pair<std::string, std::string>>(
        params_map.begin(),
        params_map.end()
    );
    auto headers = std::vector<std::pair<std::string, std::string>>(
        headers_map.begin(),
        headers_map.end()
    );

    const Result<std::string> auth_header_res = generate_auth_header(
        curl,
        key_,
        "DELETE",
        path,
        params,
        headers,
        std::vector<uint8_t>()
    );
    if (std::holds_alternative<Error>(auth_header_res)) {
        return Result<std::vector<uint8_t>>(std::get<Error>(auth_header_res));
    }

    const std::string auth_header = std::get<std::string>(auth_header_res);

    CurlSlistHandle slist;
    for (const auto& pair : headers) {
        std::stringstream header_stream;
        header_stream << pair.first << ": " << pair.second;
        const std::string header = header_stream.str();

        slist.set(curl_slist_append(slist.get(), header.c_str()));
    }

    std::stringstream auth_header_stream;
    auth_header_stream << HEADER_AUTHORIZATION << ": " << auth_header;
    const std::string auth_header_str = auth_header_stream.str();
    slist.set(curl_slist_append(slist.get(), auth_header_str.c_str()));

    std::vector<uint8_t> response;
    const std::string url =
        add_params(curl, std::get<std::string>(endpoint), params);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, slist.get());
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 0L);
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");

    if (curl_easy_perform(curl) != CURLE_OK) {
        return Result<std::vector<uint8_t>>(Error(curl.error_buffer()));
    }

    return raise_for_status(curl, response);
}
}  // namespace ul

File: java/src/main/java/com/urbanlogiq/ulsdk/ApiKeyContext.java
// Copyright (c), CommunityLogiq Software

package com.urbanlogiq.ulsdk;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.http.HttpResponse;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.URI;
import java.net.URLEncoder;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.codec.binary.Hex;
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters;
import org.bouncycastle.crypto.signers.Ed25519Signer;

public final class ApiKeyContext extends RequestContext {
    private static String canonicalizePath(String path) {
        if (path == null || path == "") {
            return "/";
        }

        if (!path.startsWith("/")) {
            return "/" + path;
        }

        return path;
    }

    private static String canonicalizeQueryString(List<Pair<String, String>> query) {
        if (query.isEmpty()) {
            return "";
        }

        query.sort((a, b) -> { return a.first().compareTo(b.first()); });

        ArrayList<String> components = new ArrayList();
        for (Pair<String, String> pair : query) {
            if (pair.first() == "X-UL-Signature") {
                continue;
            }

            String component = pair.first() + "=" + URLEncoder.encode(pair.second());
            components.add(component);
        }

        return String.join("&", components);
    }
    
    private static String canonicalizeHeaders(List<String> signedHeaders, HashMap<String, String> headers) {
        ArrayList<String> sortedSignedHeaders = new ArrayList();
        for (String header : signedHeaders) {
            sortedSignedHeaders.add(header.toLowerCase());
        }
        sortedSignedHeaders.sort(null);

        ArrayList<String> canonicalHeaderParts = new ArrayList();

        for (String v : sortedSignedHeaders) {
            try {
                String value = headers.get(v);
                canonicalHeaderParts.add(String.format("%s:%s", v, value));
            } catch (Exception e) {
                continue;
            }
        }

        return String.join("\n", canonicalHeaderParts);
    }

    private static Pair<String, String> canonicalizeRequest(String method, String path, List<Pair<String, String>> query, HashMap<String, String> headers, List<String> signedHeadersList, byte[] body) {
        String canonicalPath = canonicalizePath(path);
        String canonicalQueryString = canonicalizeQueryString(query);
        String canonicalHeaders = canonicalizeHeaders(signedHeadersList, headers);
        String signedHeaders = String.join(";", signedHeadersList);

        String s = String.format("%s\n%s\n%s\n%s\n%s\n%s",
            method.toUpperCase(),
            canonicalPath,
            canonicalQueryString,
            canonicalHeaders,
            signedHeaders,
            DigestUtils.sha256Hex(body)
        );

        return new Pair(DigestUtils.sha256Hex(s), signedHeaders);
    }

    private static String generateAuthHeader(Key key, String method, String path, List<Pair<String, String>> query, HashMap<String, String> headers, byte[] body) {
        long unixTime = System.currentTimeMillis() / 1000;
        String unixTimeString = Long.toString(unixTime);

        headers.put(HEADER_X_UL_DATE, unixTimeString);

        Pair<String, String> result = canonicalizeRequest(method, path, query, headers, SIGNED_HEADERS, body);
        String requestHash = result.first();
        String signedHeaders = result.second();

        String scope = String.format("%s/%s/%s/%s", key.userId().toString(), unixTimeString, key.region().toString(), REQUEST_TYPE);
        String signingString = String.format("%s\n%s\n%s", SIGNATURE_V1, scope, requestHash);
        byte[] signingBytes = signingString.getBytes();

        Ed25519PrivateKeyParameters params = new Ed25519PrivateKeyParameters(key.secretKey());
        Ed25519Signer signer = new Ed25519Signer();

        signer.init(true, params);
        signer.update(signingBytes, 0, signingBytes.length);
        byte[] signatureBytes = signer.generateSignature();

        String signature = Hex.encodeHexString(signatureBytes);

        return String.format(
            "%s Credential=%s/%s, SignedHeaders=%s, Signature=%s",
            SIGNATURE_V1,
            key.accessKey(),
            scope,
            signedHeaders,
            signature
        );
    }

    public ApiKeyContext(Key key, Environment environment) {
        this._key = key;
        this._environment = environment;
    }

    public Region region() {
        return this._key.region();
    }

    public Environment environment() {
        return this._environment;
    }

    public byte[] get(String path, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String host = this.getHost(this.region(), this.environment());
        String authorizationHeader = generateAuthHeader(this._key, "GET", path, params, headers, new byte[0]);

        String query = null;

        if (params.size() > 0) {
            query = canonicalizeQueryString(params);
        }

        URI uri = new URI("https", host, path, query, null);

        ArrayList<String> headerStrings = new ArrayList();
        for (String k : headers.keySet()) {
            headerStrings.add(k);
            headerStrings.add(headers.get(k));
        }
        headerStrings.add("authorization");
        headerStrings.add(authorizationHeader);

        String[] headerStringsArray = new String[headerStrings.size()];
        headerStringsArray = headerStrings.toArray(headerStringsArray);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest.Builder builder = HttpRequest.newBuilder().uri(uri);

        if (headerStringsArray.length > 0) {
            builder = builder.headers(headerStringsArray);
        }

        HttpRequest request = builder
            .GET()
            .build();

        HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());
        int statusCode = response.statusCode();

        if (statusCode < 200 || statusCode >= 300) {
            throw new FailedRequestException(response);
        }

        return response.body();
    }

    public byte[] put(String path, byte[] body, String mimetype, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        if (body == null) {
            body = new byte[0];
        }

        String host = this.getHost(this.region(), this.environment());
        String authorizationHeader = generateAuthHeader(this._key, "PUT", path, params, headers, body);

        String query = null;

        if (params.size() > 0) {
            query = canonicalizeQueryString(params);
        }

        URI uri = new URI("https", host, path, query, null);

        ArrayList<String> headerStrings = new ArrayList();
        for (String k : headers.keySet()) {
            headerStrings.add(k);
            headerStrings.add(headers.get(k));
        }
        headerStrings.add("authorization");
        headerStrings.add(authorizationHeader);

        String[] headerStringsArray = new String[headerStrings.size()];
        headerStringsArray = headerStrings.toArray(headerStringsArray);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest.Builder builder = HttpRequest.newBuilder().uri(uri);

        if (headerStringsArray.length > 0) {
            builder = builder.headers(headerStringsArray);
        }

        HttpRequest request = builder
            .PUT(HttpRequest.BodyPublishers.ofByteArray(body))
            .build();

        HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());
        int statusCode = response.statusCode();

        if (statusCode < 200 || statusCode >= 300) {
            throw new FailedRequestException(response);
        }

        return response.body();
    }

    public byte[] post(String path, byte[] body, String mimetype, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        if (body == null) {
            body = new byte[0];
        }

        String host = this.getHost(this.region(), this.environment());
        String authorizationHeader = generateAuthHeader(this._key, "POST", path, params, headers, body);

        String query = null;

        if (params.size() > 0) {
            query = canonicalizeQueryString(params);
        }

        URI uri = new URI("https", host, path, query, null);

        ArrayList<String> headerStrings = new ArrayList();
        for (String k : headers.keySet()) {
            headerStrings.add(k);
            headerStrings.add(headers.get(k));
        }
        headerStrings.add("authorization");
        headerStrings.add(authorizationHeader);

        String[] headerStringsArray = new String[headerStrings.size()];
        headerStringsArray = headerStrings.toArray(headerStringsArray);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest.Builder builder = HttpRequest.newBuilder().uri(uri);

        if (headerStringsArray.length > 0) {
            builder = builder.headers(headerStringsArray);
        }

        HttpRequest request = builder
            .POST(HttpRequest.BodyPublishers.ofByteArray(body))
            .build();

        HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());
        int statusCode = response.statusCode();

        if (statusCode < 200 || statusCode >= 300) {
            throw new FailedRequestException(response);
        }

        return response.body();
    }

    public byte[] upload(String path, List<File> files) throws URISyntaxException, IOException, InterruptedException {
        long ts = System.currentTimeMillis();
        String boundary = String.format("UL1-multipart-%d", ts);
        byte[] dash = new byte[]{'-', '-'};
        byte[] boundaryStream = boundary.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] crlf = new byte[]{'\r', '\n'};

        ByteArrayOutputStream body = new ByteArrayOutputStream();
        for (int i = 0; i < files.size(); i++) {
            File file = files.get(i);
            body.write(dash, 0, dash.length);
            body.write(boundaryStream, 0, boundaryStream.length);
            body.write(crlf, 0, crlf.length);

            String dispositionHeader = String.format("Content-Disposition: form-data; name=\"%s\"; filename=\"%s\"\r\nContent-Type: %s\r\n\r\n", file.getName(), file.getName(), file.getMimetype());
            byte[] dispositionHeaderBytes = dispositionHeader.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            body.write(dispositionHeaderBytes, 0, dispositionHeaderBytes.length);

            byte[] data = file.getData();
            body.write(data, 0, data.length);
        }

        body.write(dash, 0, dash.length);
        body.write(boundaryStream, 0, boundaryStream.length);
        body.write(dash, 0, dash.length);

        String mimetype = String.format("multipart/form-data; boundary=\"%s\"", boundary);

        List<Pair<String, String>> params = new ArrayList();
        HashMap<String, String> headers = new HashMap();

        return this.post(path, body.toByteArray(), mimetype, params, headers);
    }

    public byte[] delete(String path, List<Pair<String, String>> params, HashMap<String, String> headers) throws URISyntaxException, IOException, InterruptedException {
        String host = this.getHost(this.region(), this.environment());
        String authorizationHeader = generateAuthHeader(this._key, "DELETE", path, params, headers, new byte[0]);

        String query = null;

        if (params.size() > 0) {
            query = canonicalizeQueryString(params);
        }

        URI uri = new URI("https", host, path, query, null);

        ArrayList<String> headerStrings = new ArrayList();
        for (String k : headers.keySet()) {
            headerStrings.add(k);
            headerStrings.add(headers.get(k));
        }
        headerStrings.add("authorization");
        headerStrings.add(authorizationHeader);

        String[] headerStringsArray = new String[headerStrings.size()];
        headerStringsArray = headerStrings.toArray(headerStringsArray);

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest.Builder builder = HttpRequest.newBuilder().uri(uri);

        if (headerStringsArray.length > 0) {
            builder = builder.headers(headerStringsArray);
        }

        HttpRequest request = builder
            .DELETE()
            .build();

        HttpResponse<byte[]> response = client.send(request, HttpResponse.BodyHandlers.ofByteArray());
        int statusCode = response.statusCode();

        if (statusCode < 200 || statusCode >= 300) {
            throw new FailedRequestException(response);
        }

        return response.body();
    }

    private Key _key;
    private Environment _environment;
    private static final String REQUEST_TYPE = "ul1_request";
    private static final String SIGNATURE_V1 = "UL1-ED25519";
    private static final String HEADER_AUTHORIZATION = "authorization";
    private static final String HEADER_X_UL_DATE = "x-ul-date";
    private static final String HEADER_CONTENT_TYPE = "content-type";
    private static final List<String> SIGNED_HEADERS = Collections.unmodifiableList(Arrays.asList(HEADER_X_UL_DATE));
}
